### ISSUE TITLE
[discuss] client auth protocol for push/pull

### ISSUE THREAD

**Issue by cpuguy83** [moby/moby#38591]
In https://github.com/moby/moby/pull/34319, we are looking to add support for mirrors for private registries, which has brought up the topic of how to deal with auth.

Personally speaking, I don't want to see:

1. Auth credentials stored on the daemon side
2. Universal auth credentials for all clients
3. An even greater mess of the existing auth scenarios (Seriously, read through the code and tell me how long it takes you to figure out all that is happening).

**Comment by cpuguy83** [comment]
ping @stevvooe @dmcgowan @saschagrunert

**Comment by cpuguy83** [comment]
Just a thought on this topic...

Perahaps `docker pull <registry>/<image>` would  trigger a minor back and forth to negotiate the remote to actually pull from.... so the daemon could respond with "oh I have these mirrors configured" and the client could choose the remote(s) to try by sending the auth creds for the remote to use.

Additionally, the client could send the remote and/or the list of remotes it wants to use with the creds for them along with the initial request.... however daemon configured mirrors gets a little weird here.

It may be helpful to have an endpoint on the daemon that can assist the client in resolving a remote?

**Comment by dekkagaijin** [comment]
Having an auth negotiation between client and server would also help with the issue that `docker build` unnecessarily sends the entire universe of creds to the daemon

**Comment by thaJeztah** [comment]
@caervs PTAL

**Comment by tonistiigi** [comment]
> Having an auth negotiation between client and server would also help with the issue that docker build unnecessarily sends the entire universe of creds to the daemon

This is not true anymore with BuildKit. BuildKit uses session endpoint to call back to the client for specific credentials per host, basically like the diagram above shows but without sending requests again.

This can be extended to pull/push endpoint, mirror credentials, and short-lived tokens instead. Extra security benefit would be that the password never leaves the client machine (https://github.com/moby/moby/compare/master...simonferquel:auth-streaming actually attempted to add it long time ago).

**Comment by lajoie** [comment]
Has this work stalled (and thus stalled #34319)?

**Comment by cpuguy83** [comment]
This is really looking for contributors.

**Comment by tonistiigi** [comment]
@cpuguy83 fyi, latest development https://github.com/moby/buildkit/pull/1660

**Comment by dmcgowan** [comment]
I agree the right direction is getting tokens directly on the client. Once this is in place, it can allow token servers to generate longer lived, tightly scoped tokens which could also be used by orchestrators.

**Comment by dmcgowan** [comment]
containerd also added support for a similar flow to this in containerd 1.7. Daemon side pulling with a client callback for auth. It also supports sending `www-authenticate` header back to client and receiving back the `authorization` header to send back to registry. https://github.com/containerd/containerd/blob/main/api/types/transfer/registry.proto#L44

**Comment by systemmonkey42** [comment]
I'm interested in the work discussed here.  My personal use case is a container registry attached to a gitlab instance.

The token required to authenticate with both git and docker is created via the "create personal token" menu, and expires regularly.  For simplicity I load the token into hashicorp VAULT and use a git helper to query the vault for the credentials. (I reauthenticate with the vault occasionally, allowing a range of tools to pull their individual credentials from the vault)

I would love to have a helper mechanic which could be used to retrieve credentials, allowing for the same "vault" integration I get with other tools.

Thanks

**Comment by thaJeztah** [comment]
I know @laurazard started looking into this, and was working on a PoC, but not sure if there's something to look at already.

**Comment by laurazard** [comment]
I have a working PoC that I can push on Monday/this weekend if I have time. It's got all the essential bits working, theres just some implementation questions that I wanted to use the PoC to discuss.
